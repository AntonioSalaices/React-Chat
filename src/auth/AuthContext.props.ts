import { Func } from '@Utils/types';
import { User } from '@Models/Core';

export interface AuthError {
  statusCode: number;
  message: string;
}

export interface AuthContextProps {
  user: User | null;
  login: Func;
  logout: Func;
  loading: boolean;
  error: AuthError;
}
