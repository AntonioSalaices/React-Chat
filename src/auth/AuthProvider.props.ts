export interface AuthContextProviderProps {
  /**
   * Children node
   */
  children: JSX.Element | JSX.Element[];
}
