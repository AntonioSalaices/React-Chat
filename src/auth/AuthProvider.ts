/**
 * Date 30/06/23
 *
 * @author Antonio Salaices
 */

const token = 'cookieMethod';

const isAuthenticated = () => {
  return Boolean(token);
};

const auth = {
  isAuthenticated,
};

export { auth };
