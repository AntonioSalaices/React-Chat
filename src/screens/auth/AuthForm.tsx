import { HTMLType } from '@Constans/htmlConstants';
import { translate } from '@Translate/translate';
import { AuthFormProps } from './AuthForm.props';
import { Container, FormField } from '@Components/Core';
import { useRef, useState } from 'react';
import { FaRegEye, FaRegEyeSlash } from 'react-icons/fa';
import { usePostQuery } from '@Hooks/usePost/usePost';

const { VITE_BASE_API_URL, VITE_BASE_AUTH_URL } = import.meta.env;

const AuthForm: React.FC<AuthFormProps> = () => {
  const { responseData, loading, error, post } = usePostQuery(`${VITE_BASE_API_URL}${VITE_BASE_AUTH_URL}`);
  const [isShownPassword, setIsShownPassword] = useState<boolean>(false);
  const email = useRef<HTMLInputElement>(null);
  const password = useRef<HTMLInputElement>(null);
  console.log('response', responseData, loading, error);
  const onSubmit = (e) => {
    e.preventDefault();
    const body = {
      email: email?.current?.value,
      password: password?.current?.value,
    };
    post(body);
  };

  return (
    <Container>
      <div className="col justify-center align-center">
        <FormField
          ref={email}
          type={HTMLType.text}
          subTx="subPlaceholder.email"
          tx="form.inputs.email"
          testID="emailInput"
        />
        <FormField
          ref={password}
          type={isShownPassword ? HTMLType.text : HTMLType.password}
          tx="form.inputs.password"
          testID="passwordInput"
          right={
            isShownPassword ? (
              <FaRegEyeSlash className="pointer" onClick={() => setIsShownPassword(false)} />
            ) : (
              <FaRegEye className="pointer" onClick={() => setIsShownPassword(true)} />
            )
          }
        />
        <button className="btn-primary text-white" onClick={onSubmit}>
          {translate('buttons.signIn')}
        </button>
      </div>
    </Container>
  );
};

export default AuthForm;
