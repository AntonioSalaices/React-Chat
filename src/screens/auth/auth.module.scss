@import '../../assets/styles/palette';
@import '../../assets/styles/fonts';
@import '../../assets/styles/utilities';

.modalBackground {
  background-color: $background-color;
  width: 100vw;
  height: 100vh;
  z-index: 0;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  position: absolute;
}

.centered {
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.modal {
  z-index: 10;
  width: 50vw;
  height: 50vh;
  background-color: $white;
  border-radius: $base-border-radius;
}
