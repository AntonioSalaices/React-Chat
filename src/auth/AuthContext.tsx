import { createContext, useEffect, useState } from 'react';
import { isEmpty } from 'lodash';

import { Func } from '@Utils/types';
import { Security } from '@Constans/securityConstants';
import { User, UserKeys } from 'models/User';
import { AuthContextProviderProps } from './AuthProvider.props';

import { usePostQuery } from '@Hooks/usePost/usePost';

const { VITE_BASE_API_URL, VITE_AUTH_LOGIN_URL, VITE_AUTH_LOGOUT_URL } = import.meta.env;

interface AuthContextProps {
  user: User | null;
  login: Func;
  logout: Func;
  loading: boolean;
}

export const AuthContext = createContext<AuthContextProps>({
  user: {} as User,
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  login: () => {},
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  logout: () => {},
  loading: false,
});

export const AuthProvider: React.FC<AuthContextProviderProps> = ({ children }) => {
  const [user, setUser] = useState<User | null>(() => {
    const user = localStorage.getItem(Security.UserData);

    if (user) return JSON.parse(user);

    return null;
  });

  const {
    responseData: responseLogin,
    loading: isLoadingLogin,
    post: loginPost,
  } = usePostQuery<User, User>(`${VITE_BASE_API_URL}${VITE_AUTH_LOGIN_URL}`);

  const { post: logoutPost } = usePostQuery(`${VITE_BASE_API_URL}${VITE_AUTH_LOGOUT_URL}`);

  useEffect(() => {
    const saveResponse = () => {
      if (!isEmpty(responseLogin) && UserKeys.Email in responseLogin && UserKeys.Id in responseLogin) {
        const userData = responseLogin as User;
        setUser(userData);
        localStorage.setItem(Security.UserData, JSON.stringify(userData));
      }
    };
    saveResponse();
  }, [responseLogin]);

  const login = (credentials: User) => {
    loginPost(credentials);
  };

  const logout = () => {
    logoutPost();
    setUser(null);
    localStorage.removeItem(Security.UserData);
  };

  return (
    <AuthContext.Provider value={{ user, login, logout, loading: isLoadingLogin }}>{children}</AuthContext.Provider>
  );
};
