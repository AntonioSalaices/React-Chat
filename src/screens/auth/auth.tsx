import AuthForm from './AuthForm';
import { Modal } from '@Components/Core';
import usePortal from '@Hooks/usePortal/usePortal';
import { Func } from '@Utils/types';

interface AuthModalProps {
  isShown: boolean;
  onClose: Func;
}

export const AuthModal: React.FC<AuthModalProps> = ({ isShown, onClose }) => {
  const Portal = usePortal(document.getElementById('portal-root') as HTMLElement);

  return (
    <>
      {isShown && (
        <Portal>
          <Modal onClose={onClose}>
            <AuthForm />
          </Modal>
        </Portal>
      )}
    </>
  );
};
AuthModal.displayName = 'AuthModal';
