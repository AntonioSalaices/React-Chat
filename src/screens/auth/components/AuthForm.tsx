import { useContext, useRef, useState } from 'react';
import classNames from 'classnames';
import { FaRegEye, FaRegEyeSlash } from 'react-icons/fa';

import { color } from '@Utils/colors';

import { HTMLType, OnChangeType } from '@Constans/htmlConstants';

import { Container, FormField, Spinner } from '@Components/Core';

import { User } from 'models/User';
import { AuthFormProps } from './AuthForm.props';

import { translate } from '@Translate/translate';
import styles from './AuthForm.module.scss';
import { AuthContext } from 'auth/AuthContext';
import { useNavigate } from 'react-router-dom';

const AuthForm: React.FC<AuthFormProps> = () => {
  const navigate = useNavigate();
  const { loading, login } = useContext(AuthContext);

  const [isShownPassword, setIsShownPassword] = useState<boolean>(false);
  const [password, setPassword] = useState<string>('');
  const email = useRef<HTMLInputElement>(null);

  const onSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const body = {
      email: email?.current?.value,
      password: password,
    } as User;
    login(body);
    navigate('/dashboard');
  };

  const handlePassword = ({ target: { value } }: OnChangeType) => {
    setPassword(value);
  };

  const isSubmitButtonDisabled = !email.current?.value || !password;

  const buttonStyle = classNames('btn-primary text-white', {
    [styles.btnDisabled]: isSubmitButtonDisabled,
  });

  return (
    <Container>
      <form onSubmit={onSubmit}>
        <div className="col justify-center align-center">
          <FormField
            ref={email}
            type={HTMLType.text}
            subTx="subPlaceholder.email"
            tx="form.inputs.email"
            testID="emailInput"
          />
          <FormField
            type={isShownPassword ? HTMLType.text : HTMLType.password}
            tx="form.inputs.password"
            testID="passwordInput"
            onChange={handlePassword}
            right={
              isShownPassword ? (
                <FaRegEyeSlash className="pointer" onClick={() => setIsShownPassword(false)} />
              ) : (
                <FaRegEye className="pointer" onClick={() => setIsShownPassword(true)} />
              )
            }
          />
          <button disabled={isSubmitButtonDisabled} className={buttonStyle} onClick={onSubmit}>
            {loading ? <Spinner size={20} singleColor={color.background} /> : translate('buttons.signIn')}
          </button>
        </div>
      </form>
    </Container>
  );
};

export default AuthForm;
