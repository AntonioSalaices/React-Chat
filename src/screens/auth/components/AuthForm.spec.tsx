import { RenderResult, render } from '@testing-library/react';
import AuthForm from './AuthForm';
import { BrowserRouter } from 'react-router-dom';
import { AuthProvider } from 'auth/AuthContext';

describe('<AuthForm />', () => {
  let component: RenderResult;
  beforeEach(() => {
    component = render(
      <BrowserRouter>
        <AuthProvider>
          <AuthForm />
        </AuthProvider>
      </BrowserRouter>
    );
  });
  it('Should render AuthForm', () => {
    const { container } = component;

    expect(container).toMatchSnapshot();
  });
});
