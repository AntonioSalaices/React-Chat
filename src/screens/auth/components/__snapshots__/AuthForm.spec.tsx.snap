// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthForm /> Should render AuthForm 1`] = `
<div>
  <div
    class="container"
    id="main-container"
  >
    <form>
      <div
        class="col justify-center align-center"
      >
        <div
          class="col-11-xs col-11-sm col-5-md col-5-xl"
        >
          <div
            class="gap-2 input-content"
          >
            <label>
              [missing "en-US.form.inputs.email" translation]
              <input
                class="input-primary"
                data-testid="emailInput"
                placeholder="[missing "en-US.form.inputs.email" translation]"
                type="text"
                value=""
              />
            </label>
            <h5
              class="text-primary-light-7 mb-1"
            >
              [missing "en-US.subPlaceholder.email" translation]
            </h5>
          </div>
        </div>
        <div
          class="col-11-xs col-11-sm col-5-md col-5-xl"
        >
          <div
            class="gap-2 input-content"
          >
            <label>
              [missing "en-US.form.inputs.password" translation]
              <input
                class="input-primary"
                data-testid="passwordInput"
                placeholder="[missing "en-US.form.inputs.password" translation]"
                type="password"
                value=""
              />
            </label>
            <div
              class="input-icon-right"
            >
              <span>
                <svg
                  class="pointer"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 576 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M288 144a110.94 110.94 0 0 0-31.24 5 55.4 55.4 0 0 1 7.24 27 56 56 0 0 1-56 56 55.4 55.4 0 0 1-27-7.24A111.71 111.71 0 1 0 288 144zm284.52 97.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400c-98.65 0-189.09-55-237.93-144C98.91 167 189.34 112 288 112s189.09 55 237.93 144C477.1 345 386.66 400 288 400z"
                  />
                </svg>
              </span>
            </div>
            <h5
              class="text-primary-light-7 mb-1"
            />
          </div>
        </div>
        <button
          class="btn-primary text-white btnDisabled"
          disabled=""
        >
          [missing "en-US.buttons.signIn" translation]
        </button>
      </div>
    </form>
  </div>
</div>
`;
